src/V3EmitCFunc.h

```diff
@@ -1059,9 +1059,33 @@ class EmitCFunc VL_NOT_FINAL : public EmitCConstInit {
             putbs("(");
             iterateAndNextConstNull(nodep->condp());
             putbs(" ? ");
+            // All class types are castable to each other. If they are of different types,
+            // a compilation error will be thrown, so an explicit cast is required. Types were
+            // already checked by V3Width and dtypep of a condition operator is a type of their
+            // common base class, so both classes can be safetly casted.
+            const AstClassRefDType* const thenClassDtypep
+                = VN_CAST(nodep->thenp()->dtypep(), ClassRefDType);
+            const AstClassRefDType* const elseClassDtypep
+                = VN_CAST(nodep->elsep()->dtypep(), ClassRefDType);
+            const bool castRequired = thenClassDtypep && elseClassDtypep
+                                      && (thenClassDtypep->classp() != elseClassDtypep->classp());
+            bool thenCastRequired = false;
+            bool elseCastRequired = false;
+            std::string castStr;
+            if (castRequired) {
+                const AstClass* const commonBaseClassp
+                    = VN_AS(nodep->dtypep(), ClassRefDType)->classp();
+                if (thenClassDtypep->classp() != commonBaseClassp) thenCastRequired = true;
+                if (elseClassDtypep->classp() != commonBaseClassp) elseCastRequired = true;
+                castStr = "static_cast<" + nodep->dtypep()->cType("", false, false) + ">(";
+            }
```

Comment:
I think instead V3Cast should insert an AstCCast into the tree, then the normal Emit code will handle it.  Unfortunately AstCCast doesn't seem to currently allow arbitrary types, so I'd suggest add a string parameter to AstCCast to replace the current integer size.

Reply:
I added this cast here, because I think that the condition expression is the only expression in which that problem may occur and is directly related to the generation of c++ code, because this cast is needed only to satisfy c++ type control.

I didn't notice that there is a special stage for inserting such casts. I will move it.

Reply:
I didn't have to add a string parameter. I used dtypep instead.